### DISCUSSION TITLE
Team B - Course 2

### DISCUSSION THREAD

**Discussion by mindgitrwx** [lablup/contribution-academy-2021#37] [Show and tell]
Q1. 아래 저장소의 maintainer 를 찾고, 그 근거를 찾을 수 있는 git 명령어를 올려주세요.

    Backend.AI-manager
    Backend.AI-agent
    Backend.AI-common
    Backend.AI-client-py
    Backend.AI-webserver
    Backend.AI-webui
    Backend.AI-storage-proxy

Q2. Backend.AI project에 Write할 수 있는 권한이 있는 John과 권한이 없는 Jane 이 있습니다.
이 때, Jane 이 구현한 기능이 main branch에 적용되기 위한 단계는 John 이 기능을 main branch에
merge 할 때와 어떤 다른 점이 있고, 그 이유는 무엇인지 두 문장 이내로 간단하게 적어주세요.

**Comment by mindgitrwx** [comment]
Q1의 경우, 위 모든 repository 들을 한 폴더에 클론한 뒤 다음 bash script 파일을 만들어 실행시켰습니다.

```bash
#!/bin/bash
for d in ./*/ ; 
do 
    (cd "$d" && basename "$PWD" &&  git log | grep Author | sort | uniq -c | sort -n -r | head -n 1); 
done

```
순서대로 다음과 같은 결과가 나옵니다. 

Backend.ai-agent
   1142 Author: Joongi Kim <joongi@lablup.com>
Backend.ai-client-py
    728 Author: Joongi Kim <joongi@lablup.com>
Backend.ai-common
    506 Author: Joongi Kim <joongi@lablup.com>
Backend.ai-manager
   1518 Author: Joongi Kim <joongi@lablup.com>
Backend.ai-storage-proxy
     80 Author: Joongi Kim <joongi@lablup.com>
Backend.ai-webserver
    113 Author: Jeongkyu Shin <inureyes@gmail.com>
Backend.ai-webui
   3993 Author: Jeongkyu Shin <inureyes@gmail.com>

****2번째 라인의 for 문은 현재 폴더의 서브디렉토리 이름을 구합니다.****

****4번째 라인의 cd로 각 디렉토리로 이동합니다.****

****basename "$PWD 는 이동한 폴더의 이름을 출력합니다.****

****git log~부터 -r까지 힌트에 주어진 스크립트와 동일합니다****

****head -n 1 은 힌트에 주어진 스크립트를 실행했을 시 출력되는 텍스트에서 가장 윗 줄을 잘라 출력합니다.****

더 확장성 있는 코드도 나올 수 있습니다

Q2. John은 개인 혼자서도 자신의 remote repository 에 merge하는 과정에서 별도의 승인 과정이 필요하지 않지만, Jane 의 경우에는 자신의 개인 repository 에서 fork 하고 개발을 진행한 뒤 pull request해서 승인받는 과정이 필요합니다.  이러한 절차가 없는 경우를 가정하면, Backend.AI 코드의 전체를 누가 트롤링하여 코드를 엉망으로 고치는 일이 생길 수 있습니다.

**Comment by AlchemistX** [comment]
## Q1 : 아래 저장소의 maintainer 를 찾고, 그 근거를 찾을 수 있는 git 명령어를 올려주세요.
코드의 커밋 횟수가 많은 사람이 repo의 maintainer 라는 것을 의미하지는 않는 것 같지만, 질문의 의도가 그러한 듯 합니다.
각각의 저장소에서 커밋 횟수가 가장 많은 사람을 찾아보았습니다. 커밋 횟수가 가장 많은 분과 그다음 많은 분 사이의 차이가 큰 것을 알 수 있습니다.
그리고, 프로젝트를 세 분의 개발자가 하드캐리 하시는것으로 보입니다.

### https://github.com/lablup/backend.ai-manager
![manager](https://user-images.githubusercontent.com/1624251/129012078-17741d15-6055-46b7-bf53-0ef02f0667f0.png)

### https://github.com/lablup/backend.ai-agent
![agent](https://user-images.githubusercontent.com/1624251/129012162-f67f5741-2e98-424f-99e1-0742958cc164.png)

### https://github.com/lablup/backend.ai-common
![common](https://user-images.githubusercontent.com/1624251/129012196-761763c9-a926-4940-a119-6b9e355b03a2.png)

### https://github.com/lablup/backend.ai-client-py
![client-py](https://user-images.githubusercontent.com/1624251/129012422-f3590e5b-50df-49f9-8cea-f891f86b3de3.png)

### https://github.com/lablup/webserver
![webserver](https://user-images.githubusercontent.com/1624251/129012480-c8bac35e-34c8-4d50-8045-995159477c43.png)

### https://github.com/lablup/webui
![webui](https://user-images.githubusercontent.com/1624251/129012529-4e27bc91-8d4e-486b-9937-f16cfba18ee1.png)

### https://github.com/lablup/storage-proxy
![storage-proxy](https://user-images.githubusercontent.com/1624251/129012585-1e5b22ac-5298-4000-a7d8-60fc27426340.png)

## Q2 : Backend.AI project에 Write할 수 있는 권한이 있는 John과 권한이 없는 Jane 이 있습니다. 이 때, Jane 이 구현한 기능이 main branch에 적용되기 위한 단계는 John 이 기능을 main branch에 merge 할 때와 어떤 다른 점이 있고, 그 이유는 무엇인지 두 문장 이내로 간단하게 적어주세요.

### Jane에 잠깐 빙의 되어 상황의 맥락을 가정해 봅니다.
Jane은 upstream의 구현을 사용 중 불편한 점을 개선 하여 patch 를 생성하였고, 이것이 upstream 에 반영되면 앞으로 Backend.AI project 의 업데이트가 발생 할 때 마다 이 patch를 rebase 하지 않아도 될 것이며 이 개선점이 upstream에도 이로울 것이라 판단 합니다.

### 이런 맥락에서 Jane 이 해야 할 일은 다음과 같습니다.
Jane은 upstream에 자신의 patch를 반영하기 위하여 upstream의 repo를 자신의 계정에 fork 한 뒤 자신의 patch를 여기에 push 한 뒤 이 patch의 반영 요청 (Pull Request) 를 작성하여 upstream의 리뷰를 요청 합니다.

### 이 과정은 Backend.AI 와 Jane 모두에게 다음과 같은 이익이 있을겁니다.
서로의 필요에 의해 작성된 patch가 Backend.AI의 방향에 맞고 코드 품질이 적절한지 여러 이해관계자들의 검토 과정을 거치며 project가 건강하게 성장 합니다. Jane 역시 많은 개발자들과 소통하며 건강하게 성장할 수 있습니다.

**Comment by TeusEE** [comment]
**Q1. 아래 저장소의 maintainer 를 찾고, 그 근거를 찾을 수 있는 git 명령어를 올려주세요.**

Backend.AI-manager
Backend.AI-agent
Backend.AI-common
Backend.AI-client-py
Backend.AI-webserver
Backend.AI-webui
Backend.AI-storage-proxy

```
xodn@DESKTOP-ICxodnHaswell:~/test$ ls
backend.ai-agent      backend.ai-common   backend.ai-storage-proxy  backend.ai-webui
backend.ai-client-py  backend.ai-manager  backend.ai-webserver
xodn@DESKTOP-ICxodnHaswell:~/test$ cd backend.ai-agent
xodn@DESKTOP-ICxodnHaswell:~/test/backend.ai-agent$ git log | grep Author | sort | uniq -c | sort -n -r | head -n 1
   1142 Author: Joongi Kim <joongi@lablup.com>
xodn@DESKTOP-ICxodnHaswell:~/test/backend.ai-agent$ cd ..
xodn@DESKTOP-ICxodnHaswell:~/test$ cd backend.ai-common
xodn@DESKTOP-ICxodnHaswell:~/test/backend.ai-common$ git log | grep Author | sort | uniq -c | sort -n -r | head -n 1
    506 Author: Joongi Kim <joongi@lablup.com>
xodn@DESKTOP-ICxodnHaswell:~/test/backend.ai-common$ cd ..
xodn@DESKTOP-ICxodnHaswell:~/test$ cd backend.ai-storage-proxy
xodn@DESKTOP-ICxodnHaswell:~/test/backend.ai-storage-proxy$ git log | grep Author | sort | uniq -c | sort -n -r | head -n 1
     80 Author: Joongi Kim <joongi@lablup.com>
xodn@DESKTOP-ICxodnHaswell:~/test/backend.ai-storage-proxy$ cd ..
xodn@DESKTOP-ICxodnHaswell:~/test$ cd backend.ai-webui
xodn@DESKTOP-ICxodnHaswell:~/test/backend.ai-webui$ git log | grep Author | sort | uniq -c | sort -n -r | head -n 1
   3993 Author: Jeongkyu Shin <inureyes@gmail.com>
xodn@DESKTOP-ICxodnHaswell:~/test/backend.ai-webui$ cd ..
xodn@DESKTOP-ICxodnHaswell:~/test$ cd backend.ai-client-py
xodn@DESKTOP-ICxodnHaswell:~/test/backend.ai-client-py$ git log | grep Author | sort | uniq -c | sort -n -r | head -n 1
    728 Author: Joongi Kim <joongi@lablup.com>
xodn@DESKTOP-ICxodnHaswell:~/test/backend.ai-client-py$ cd ..
xodn@DESKTOP-ICxodnHaswell:~/test$ cd backend.ai-manager
xodn@DESKTOP-ICxodnHaswell:~/test/backend.ai-manager$ git log | grep Author | sort | uniq -c | sort -n -r | head -n 1
   1518 Author: Joongi Kim <joongi@lablup.com>
xodn@DESKTOP-ICxodnHaswell:~/test/backend.ai-manager$ cd ..
xodn@DESKTOP-ICxodnHaswell:~/test$ cd backend.ai-webserver
xodn@DESKTOP-ICxodnHaswell:~/test/backend.ai-webserver$ git log | grep Author | sort | uniq -c | sort -n -r | head -n 1
    113 Author: Jeongkyu Shin <inureyes@gmail.com>
```

WSL 우분투에서 test folder를 생성하여 7개의 repo를 clone한 다음 각각 의 repo를 이동하면서 가장 contribution을 많이 한 maintainer를 확인하였고, 리스트는 아래와 같습니다.
backend.ai-agent : Joongi Kim <joongi@lablup.com>, 1142회
backend.ai-common : Joongi Kim <joongi@lablup.com>, 506회
backend.ai-storage-proxy : Joongi Kim <joongi@lablup.com>, 80회
backend.ai-webui : Jeongkyu Shin <inureyes@gmail.com>, 3993회
backend.ai-client-py : Joongi Kim <joongi@lablup.com>, 728회
backend.ai-manager : Joongi Kim <joongi@lablup.com>, 1518회
backend.ai-webserver : Jeongkyu Shin <inureyes@gmail.com>, 113회

"git log | grep Author | sort | uniq -c | sort -n -r | head -n 1" 를 사용해서 git log에서 Author를 기준으로 count하여 정열하였을 때 가장 높은 순위를 추출하였습니다.

**Q2. Backend.AI project에 Write할 수 있는 권한이 있는 John과 권한이 없는 Jane 이 있습니다.
이 때, Jane 이 구현한 기능이 main branch에 적용되기 위한 단계는 John 이 기능을 main branch에
merge 할 때와 어떤 다른 점이 있고, 그 이유는 무엇인지 두 문장 이내로 간단하게 적어주세요.**

=>권한이 없는 Jane은 변경사항을 Forked된 repo에 push한 후 PR을 진행해야지만 main branch에 적용이 가능하다. 하지만 권한이 있는 John은 별도의 Fork없이 바로 repo를 local에서 수정한 뒤 push를 통해서 PR의 과정 없이 merge가 가능하다. 하지만 이러한 경우는 오픈소스의 관리 측면에서 부적절하다

**Comment by min1826** [comment]
### Q1. 아래 저장소의 maintainer 를 찾고, 그 근거를 찾을 수 있는 git 명령어를 올려주세요.

* **각 저장소의 maintainer(=전 기간에 걸쳐 해당 저장소에 가장 코드 기여가 많은 contributor)**
Backend.AI-manager : Joongi Kim
Backend.AI-agent : Joongi Kim
Backend.AI-common : Joongi Kim
Backend.AI-client-py : Joongi Kim
Backend.AI-webserver : Jeongkyu Shin
Backend.AI-webui : Jeongkyu Shin
Backend.AI-storage-proxy : Joongi Kim

* **git 명령어**
`git log | grep Author | sort | uniq -c | sort -n -r | head -n 1`

* **실행화면(일부)**
```bash
MinKyeong@MIN-HOME MINGW64 /d/backend_AI/course2/backend.ai-manager (main)
$ git log | grep Author | sort | uniq -c | sort -n -r | head -n 1
   1518 Author: Joongi Kim <joongi@lablup.com>

MinKyeong@MIN-HOME MINGW64 /d/backend_AI/course2/backend.ai-agent (main)
$ git log | grep Author | sort | uniq -c | sort -n -r | head -n 1
   1142 Author: Joongi Kim <joongi@lablup.com>

MinKyeong@MIN-HOME MINGW64 /d/backend_AI/course2/backend.ai-common (main)
$ git log | grep Author | sort | uniq -c | sort -n -r | head -n 1
    506 Author: Joongi Kim <joongi@lablup.com>

MinKyeong@MIN-HOME MINGW64 /d/backend_AI/course2/backend.ai-client-py (main)
$ git log | grep Author | sort | uniq -c | sort -n -r | head -n 1
    728 Author: Joongi Kim <joongi@lablup.com>

MinKyeong@MIN-HOME MINGW64 /d/backend_AI/course2/backend.ai-webserver (main)
$ git log | grep Author | sort | uniq -c | sort -n -r | head -n 1
    113 Author: Jeongkyu Shin <inureyes@gmail.com>

MinKyeong@MIN-HOME MINGW64 /d/backend_AI/course2/backend.ai-webui (main)
$ git log | grep Author | sort | uniq -c | sort -n -r | head -n 1
   3993 Author: Jeongkyu Shin <inureyes@gmail.com>

MinKyeong@MIN-HOME MINGW64 /d/backend_AI/course2/backend.ai-storage-proxy (main)
$ git log | grep Author | sort | uniq -c | sort -n -r | head -n 1
     80 Author: Joongi Kim <joongi@lablup.com>
```

### Q2. Backend.AI project에 Write할 수 있는 권한이 있는 John과 권한이 없는 Jane 이 있습니다. 이 때, Jane 이 구현한 기능이 main branch에 적용되기 위한 단계는 John 이 기능을 main branch에 merge 할 때와 어떤 다른 점이 있고, 그 이유는 무엇인지 두 문장 이내로 간단하게 적어주세요.
write할 수 있는 권한이 있는 John은 원본 저장소 안에서 branch를 만들어서 작업한 후 main branch에 merge 하지만, write할 수 있는 권한이 없는 Jane은 fork한 후 Jane의 GitHub account에서 branch 만들어서 작업하고 Pull Request 해야 합니다. 차이가 있는 이유는 write할 수 있는 권한의 유무 때문입니다.
